### PR TITLE
Myelin components for training

### PR DESCRIPTION
**Pull request by ringgaard**
Myelin components and extensions for using Myelin at training-time. Major changes include:

* Myelin flows now supports global learnable parameters as well as global constants. The learned weights can be saved back into a flow after training is complete.

* Myelin now has a gradient module which can be used for building a gradient function for a function. These can be used for reverse-mode auto differentiation for computing gradients for gradient descent during learning. Components for constructing the gradient for an operation can be registered in the Myelin component library.

* A module for cross entropy loss computation can be used for computing the loss for multi-class classification problems. This also computes the gradient of the loss which can be propagated back through the gradient functions.

* A stochastic gradient descent optimizer for applying the gradients to the global parameters to the model has been added. This also supports norm clipping and regularization. There is also an optimizer based on the Adam update rule.

* Support for assign operations. These are used for parameter updates. The expression generator has been extended to also support assignment. The element-wise index generator has been refactored so locators and iterators can be shared between input and output tensors. The expression generator now supports broadcast loads to expressions with scalar inputs will not force single-element mode. The Express module now supports both register-based input and output variables. It can now also do hoisting of loop-invariant ops out of the body loading constant and scalars as  well as operations on these can be computed outside the loop. High-level methods for expression optimization and code generation has been added. Square root and Relu gradient operations have been added to Express. The Negate op has been renamed to Neg to be compatible with TF. Unused inputs to calculate ops are now eliminated. Inputs and outputs are now checking against overlapping live ranges in the calculation before sharing.

* BiLSTM module for building and computing bi-directional LSTM activations and gradients.

* New kernels:
  - Softmax kernel for cross entropy loss
  - OneHot kernel for generating one-hot vectors
  - Slice kernel for extracting tensor slices
  - GatherSum, GatherMax, and GatherAvg kernels for embedding lookups
  - ScatterAdd for sparse updating of embedding matrix
  - Norm for computing L2 norm of tensor
  - AVXFltDotProduct for optimized dot product
  - AVXFltAssignAddOuter for optimized outer product with assignment
  - Reference op for accessing parameters in other cells of the network

* New transformations for combined ops:
  - Transformation of MatMul->Add->Assign into combined op.
  - Transformation of Add->Scatter->Assign into combined op.
  - Transformation of Mul->Add->Scatter into combined op.
  - Transform Add(x, Neg(y)) to Sub(x, y).
  - Propagation of refs through Reshape ops.
  - Merge transpose into MatMul attributes.

* The myelin::Builder has been renamed to myelin::FlowBuilder to avoid name  clash with sling::Builder for frames. The flow builder has been almost  completely rewritten and I have added more ops to the builder. There are now also methods for building LSTMs and FF layers. The flow builder now produces flows with explicit type and shape information. There is now also a Scope class for the flow builder for nested name spaces.

* Flow file format version 5 with flags field for the flow artifacts. The old versions are still supported. The in, out, and ref fields have been turned  into bits in the flags field. The attribute interface has been refactored to make the code simpler for artifacts with attributes (less code duplication).

* Minor changes:
  - The flow is passed together with op for type inference components.
  - Linear indexing into tensors to traverse all elements in standard order. The makes it easier to do element-wise operations, e.g. initialization and copying.
  - Linked tensors are now inserted into a circular list, so layout constraints can now be enforced transitively.
  - Global tensors can now be output as strings for debugging.
  - The Connector class has been removed and channels are now created based on the format of a tensor.
  - Global profiling model where each cell holds the profile summary.
  - Global FLOP counter for measuring FLOPS without having to run in profiling mode. Code is generated by the compiler to add the number of floating point operations to a global counter each time a cell function is executed.
  - Myelin linker is now notified both before and after step generation.
  - ELF linker now supports putting parameter data into the BSS section so the weights can be loaded at runtime. Useful for AOT compilation.
  - The order of step processing has been changed so all steps are now created before the Kernel::Support methods are used for determining the kernel for a particular step. Also, the Kernel::Adjust method is called immediately after a kernel has been selected.
  - The r15 register is used in debug mode instead of rbp as the data instance register so the stack frames can be traversed correctly during debugging.
  - DOT graph generator now outputs more descriptive tool tips for ops.
  - Removed legacy kernels where functionality has been replaced by the  expression generator.
  - Reading scalar and vector values from tensor.
  - Network::GetParameter() now CHECK if tensor name exists.
  - Zeroing of elements in channel.
  - The data for a tensors is initialized to the value from the flow so the value can be used before tensor data allocation.
  - The TypeTraits now include the corresponding C type.
  - Shape transpose now returns the transposed shape.
  - Report number of CPU cores in profile report.
  - Detection of AVX512 CPU features.
  - Set the FTZ and DAZ CPU status flags to prevent slowdown on denormal numbers.
  - Improved checking for vzeroupper.

### CODE REVIEW

**Comment by ringgaard** [comment]
I have made a [simple trainer](https://github.com/ringgaard/sling/blob/dev/sling/myelin/tests/lextagger.cc) for a POS tagger as an example of using the Myelin learning components. 

NB: It uses the [lexical encoder](https://github.com/ringgaard/sling/blob/dev/sling/nlp/document/lexical-features.h) which has not been checked into the caspar branch yet.

File: sling/myelin/kernel/arithmetic.cc [hunk 1]
```diff
@@ -406,6 +559,36 @@ class ExpressionTransformer : public Transformer {
...
+    for (Flow::Operation *op : flow->ops()) {
+      bool calculate = op->type == "Calculate";
+      bool assign = op->type == "Assign";
+      if (calculate || assign) {
+        Express expr;
+        InitExpression(op, &expr, false);
+        for (int i = 0; i < op->inputs.size(); ++i) {
+          if (expr.Lookup(Express::INPUT, i) == nullptr &&
+              expr.Lookup(Express::CONST, i) == nullptr) {
+            if (assign &&  i == 0) continue;
```
Comment (rahul1980) [inline sling/myelin/kernel/arithmetic.cc]: extra space after &&

File: sling/myelin/builder.cc [hunk 2]
```diff
@@ -17,39 +17,264 @@
...
+  }
+}
+
+Scope::~Scope() {
+  CHECK(root_->current_ == this);
+  root_->current_ = parent_;
+}
+
+string Scope::OpName(const string &op) {
+  string name = current_->name_;
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/myelin/builder.cc]: This is by-design. We want the prefix for the inner-most scope to be used. This allows you to use the scopes in a Python-like with fashion, e.g.:
```
FlowBuilder tf(flow, "myfunc");

{ 
  Scope subfunc1(&flow, "sf1");
  tf.Add(x,y);
}

{
  Scope subfunc2(&flow, "sf2");
  tf.Add(x,y);
}
```

I have made OpName() protected so outside users are not tempted to use this method.

File: sling/myelin/builder.cc [hunk 3]
```diff
@@ -17,39 +17,264 @@
...
+  if (num > 0) {
+    name.push_back('_');
+    name.append(std::to_string(num));
+  }
+  return name;
+}
+
+Flow::Variable *FlowBuilder::Var(const string &name, Type type,
+                                 const Shape &shape) {
+  return flow_->AddVariable(func_->name + "/" + name, type, shape);
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/myelin/builder.cc]: Yes. Fixed.

File: sling/myelin/builder.cc [hunk 4]
```diff
@@ -17,39 +17,264 @@
...
+    v = MatMul(v, W);
+    v->type = type;
+    v->shape = {1, width};
+
+    // Optionally add bias.
+    if (bias) {
+      auto *b = Parameter("b" + std::to_string(l), type, {1, width});
+      v = Add(v, b);
+      v->type = type;
+      v->shape = {1, width};
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/myelin/builder.cc]: You're right. I made FFLayers() before adding type and shape inference to the builder. I have removed these, and the code looks better now.

File: sling/myelin/builder.cc [hunk 5]
```diff
@@ -17,39 +17,264 @@
...
+    }
+
+    // Make hidden layer a reference.
+    if (l == hidden) {
+      v = Name(Identity(v), "hidden");
+      v->type = type;
+      v->shape = {1, width};
+      v->set_in()->set_out()->set_ref();
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/myelin/builder.cc]: By setting in, out, and ref on the hidden layout output, we can use this as a recurrence. By setting in and out, we ensure that this tensor is not optimized away by Myelin, and the ref allows you to store the hidden activations in a channel.

File: sling/myelin/builder.cc [hunk 6]
```diff
@@ -17,39 +17,264 @@
...
+
+  auto *x2c = Parameter("x2c", type, {input_dim, size});
+  auto *h2c = Parameter("h2c", type, {size, size});
+  auto *bc = Parameter("bc", type, {1, size});
+
+  // Channels -- h_in, c_in = h_{t-1}, c_{t-1}
+  auto *h_in = Placeholder("h_in", type, {1, size}, true);
+  auto *c_in = Placeholder("c_in", type, {1, size}, true);
+
+  // Input -- i_t = sigmoid(x_t * x2i + h_in * h2i + c_in * c2i)
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/myelin/builder.cc]: Fixed.

File: sling/myelin/express.h [hunk 7]
```diff
@@ -155,10 +157,17 @@ class Express {
     // Temporary variables and register-based inputs are in registers.
     bool IsRegister() const { return type == TEMP || type == REGISTER; }
 
+    // Check if live range for variable overlaps with another variable.
+    bool overlaps(Var *other) const {
+      return first->index < other->last->index &&
+             other->first->index < last->index;
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/myelin/express.h]: I believe this is a sufficient condition for checking for overlap. FWIW [Stack Overflow](https://stackoverflow.com/questions/3269434/whats-the-most-efficient-way-to-test-two-integer-ranges-for-overlap) seems to agree with me under the assumption that first <= last, which is true in this case.

File: sling/myelin/compute.h [hunk 8]
```diff
@@ -396,8 +428,8 @@ class Tensor {
   // are not stored on the host.
   size_t offset() const { return offset_; }
 
-  // Offset in device data instance block. Return -1 for constants and tensors
-  // that are not stored on the device.
+  // Offset in device data instance block. Return -1 for tensors
+  // that are not not stored in the instance block on the device.
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/myelin/compute.h]: Fixed

File: sling/myelin/compute.cc [hunk 9]
```diff
@@ -774,12 +839,42 @@ Network::~Network() {
...
+      var->data = dst;
+
+      // Copy elements one at a time.
+      for (int i = 0; i < elements; ++i) {
+        size_t offset = tensor->LinearOffset(i);
+        memcpy(dst, src + offset, element_size);
+        dst += element_size;
+      }
+    }
+    var->clear_learnable();
```
[the remark this answers is not in the document]
Comment (ringgaard) [reply sling/myelin/compute.cc]: By set the data for the variable and clearing the learnable flag, we are turning it into a constant.

File: sling/myelin/kernel/arithmetic.cc (hunk 1 above)
Comment (ringgaard) [reply sling/myelin/kernel/arithmetic.cc]: Fixed
